modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/domain/tokenlock/block/TokenLockBlockAcceptanceLogic.scala

```diff
@@ -0,0 +1,198 @@
+package io.constellationnetwork.node.shared.domain.tokenlock.block
+
+import cats.data.{EitherT, NonEmptyList}
+import cats.effect.Async
+import cats.syntax.all._
+
+import io.constellationnetwork.node.shared.domain.tokenlock.TokenLockChainValidator.TokenLockNel
+import io.constellationnetwork.schema.address.Address
+import io.constellationnetwork.schema.balance.{Amount, Balance, BalanceArithmeticError}
+import io.constellationnetwork.schema.tokenLock.{TokenLockBlock, TokenLockReference}
+import io.constellationnetwork.security.signature.Signed
+import io.constellationnetwork.security.{Hasher, SecurityProvider}
+
+trait TokenLockBlockAcceptanceLogic[F[_]] {
+  def acceptBlock(
+    block: Signed[TokenLockBlock],
+    txChains: Map[Address, TokenLockNel],
+    context: TokenLockBlockAcceptanceContext[F],
+    contextUpdate: TokenLockBlockAcceptanceContextUpdate
+  )(implicit hasher: Hasher[F]): EitherT[F, TokenLockBlockNotAcceptedReason, TokenLockBlockAcceptanceContextUpdate]
+
+}
+
+object TokenLockBlockAcceptanceLogic {
+
+  def make[F[_]: Async: SecurityProvider]: TokenLockBlockAcceptanceLogic[F] =
+    new TokenLockBlockAcceptanceLogic[F] {
+
+      def acceptBlock(
+        signedBlock: Signed[TokenLockBlock],
+        txChains: Map[Address, TokenLockNel],
+        context: TokenLockBlockAcceptanceContext[F],
+        contextUpdate: TokenLockBlockAcceptanceContextUpdate
+      )(implicit hasher: Hasher[F]): EitherT[F, TokenLockBlockNotAcceptedReason, TokenLockBlockAcceptanceContextUpdate] =
+        for {
+          _ <- processSignatures(signedBlock, context)
+          contextUpdate1 <- processLastTxRefs(txChains, context, contextUpdate)
+          contextUpdate2 <- processBalances(signedBlock, context, contextUpdate1)
+        } yield contextUpdate2
+
+      def processLastTxRefs(
+        txChains: Map[Address, TokenLockNel],
+        context: TokenLockBlockAcceptanceContext[F],
+        contextUpdate: TokenLockBlockAcceptanceContextUpdate
+      )(implicit hasher: Hasher[F]): EitherT[F, TokenLockBlockNotAcceptedReason, TokenLockBlockAcceptanceContextUpdate] =
+        txChains.toList
+          .foldLeft((contextUpdate.lastTokenLocksRefs, none[TokenLockBlockAwaitReason]).asRight[RejectedTokenLock].toEitherT[F]) {
+            (acc, tup) =>
```

Comment:
You can directly pattern match `tup`
```
(acc, (address, txChain)) => 
```

So there is no need to `val (address, txChain) = tup`

Reply:
It's failing for me using this pattern matching 🤔

Reply:
```
case (acc, (address, txChain)) => 
```

probably `case` is missing

Reply:
Yeah, exactly 👍🏼